Update uhdmastfrontend.cc to use changed SynthSubset constructor

@rkapuscik , please update: uhdmastfrontend.cc: In member function 'virtual Yosys::AST::AstNode* Yosys::UhdmAstFrontend::parse(std::string)':
02:42:28 | uhdmastfrontend.cc:47:120: error: no matching function for call to 'UHDM::SynthSubset::SynthSubset(UHDM::Serializer*, std::set<const UHDM::BaseClass*>&, bool)'
02:42:28 |    47 |         UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false);
02:42:28 |       |                                                                                                                        ^
02:42:28 | In file included from uhdmcommonfrontend.h:23,
02:42:28 |                  from uhdmastfrontend.cc:20:
02:42:28 | /root/Surelog/Surelog/yosys-uhdm-plugin-integration/image/include/uhdm/SynthSubset.h:39:3: note: candidate: 'UHDM::SynthSubset::SynthSubset(UHDM::Serializer*, std::set<const UHDM::BaseClass*>&, bool, bool)'
02:42:28 |    39 |   SynthSubset(Serializer* serializer,
02:42:28 |       |   ^~~~~~~~~~~
02:42:28 | /root/Surelog/Surelog/yosys-uhdm-plugin-integration/image/include/uhdm/SynthSubset.h:39:3: note:   candidate expects 4 arguments, 3 provided
02:42:28 | /root/Surelog/Surelog/yosys-uhdm-plugin-integration/image/include/uhdm/SynthSubset.h:37:7: note: candidate: 'UHDM::SynthSubset::SynthSubset(const UHDM::SynthSubset&)'
02:42:28 |    37 | class SynthSubset final : public VpiListener {
02:42:28 |       |       ^~~~~~~~~~~

--- a/systemverilog-plugin/uhdmastfrontend.cc
+++ b/systemverilog-plugin/uhdmastfrontend.cc
@@ -44,7 +44,8 @@ struct UhdmAstFrontend : public UhdmCommonFrontend {
         UHDM::Serializer serializer;
 
         std::vector<vpiHandle> restoredDesigns = serializer.Restore(filename);
-        UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false, true);
+        UHDM::SynthSubset *synthSubset =
+          make_new_object_with_optional_extra_true_arg<UHDM::SynthSubset>(&serializer, this->shared.nonSynthesizableObjects, false);
         synthSubset->listenDesigns(restoredDesigns);
         delete synthSubset;
         if (this->shared.debug_flag || !this->report_directory.empty()) {

--- a/systemverilog-plugin/uhdmastfrontend.cc
+++ b/systemverilog-plugin/uhdmastfrontend.cc
@@ -44,7 +44,7 @@ struct UhdmAstFrontend : public UhdmCommonFrontend {
         UHDM::Serializer serializer;
 
         std::vector<vpiHandle> restoredDesigns = serializer.Restore(filename);
-        UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false);
+        UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false, true);
         synthSubset->listenDesigns(restoredDesigns);
         delete synthSubset;
         if (this->shared.debug_flag || !this->report_directory.empty()) {

--- a/systemverilog-plugin/uhdmcommonfrontend.h
+++ b/systemverilog-plugin/uhdmcommonfrontend.h
@@ -23,9 +23,27 @@
 #include "uhdm/SynthSubset.h"
 #include "uhdm/VpiListener.h"
 #include <string>
+#include <type_traits>
 #include <vector>
 
 YOSYS_NAMESPACE_BEGIN
+
+// FIXME (mglb): temporary fix to support UHDM both before and after the following change:
+// https://github.com/chipsalliance/UHDM/commit/d78d094448bd94926644e48adea4df293b82f101
+// The commit introducing this code should to be reverted after Surelog is bumped to recent versions in all our repositories.
+template <typename ObjT, typename... ArgN, std::enable_if_t<std::is_constructible_v<ObjT, ArgN...>, bool> = true>
+static inline ObjT *make_new_object_with_optional_extra_true_arg(ArgN &&... arg_n)
+{
+    // Older UHDM version
+    return new ObjT(std::forward<ArgN>(arg_n)...);
+}
+
+template <typename ObjT, typename... ArgN, std::enable_if_t<!std::is_constructible_v<ObjT, ArgN...>, bool> = true>
+static inline ObjT *make_new_object_with_optional_extra_true_arg(ArgN &&... arg_n)
+{
+    // Newer UHDM version
+    return new ObjT(std::forward<ArgN>(arg_n)..., true);
+}
 
 struct UhdmCommonFrontend : public Frontend {
     UhdmAstShared shared;

--- a/systemverilog-plugin/uhdmsurelogastfrontend.cc
+++ b/systemverilog-plugin/uhdmsurelogastfrontend.cc
@@ -149,7 +149,7 @@ struct UhdmSurelogAstFrontend : public UhdmCommonFrontend {
         // Should be called 1. for normal flow 2. after finishing with `-link`
         if (!this->shared.defer) {
             UHDM::Serializer serializer;
-            UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false);
+            UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false, true);
             synthSubset->listenDesigns(uhdm_design);
             delete synthSubset;
         }

--- a/systemverilog-plugin/uhdmsurelogastfrontend.cc
+++ b/systemverilog-plugin/uhdmsurelogastfrontend.cc
@@ -149,7 +149,8 @@ struct UhdmSurelogAstFrontend : public UhdmCommonFrontend {
         // Should be called 1. for normal flow 2. after finishing with `-link`
         if (!this->shared.defer) {
             UHDM::Serializer serializer;
-            UHDM::SynthSubset *synthSubset = new UHDM::SynthSubset(&serializer, this->shared.nonSynthesizableObjects, false, true);
+            UHDM::SynthSubset *synthSubset =
+              make_new_object_with_optional_extra_true_arg<UHDM::SynthSubset>(&serializer, this->shared.nonSynthesizableObjects, false);
             synthSubset->listenDesigns(uhdm_design);
             delete synthSubset;
         }

